File: src/main/java/com/cxxcxx/zinecraft/core/dimension/StarGateFeature.java
package com.cxxcxx.zinecraft.core.dimension;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction.Axis;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.levelgen.Heightmap.Types;
import net.minecraft.world.level.levelgen.feature.Feature;
import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext;
import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
import org.jetbrains.annotations.NotNull;

public final class StarGateFeature extends Feature<NoneFeatureConfiguration> {
  public StarGateFeature() {
    super(NoneFeatureConfiguration.CODEC);
  }

  public boolean place(@NotNull FeaturePlaceContext<NoneFeatureConfiguration> context) {
    WorldGenLevel worldGenLevel = context.level();
    if (!java.util.Objects.equals(worldGenLevel.getLevel().dimension(), Level.OVERWORLD)) {
      return false;
    }

    BlockPos blockPos = worldGenLevel.getHeightmapPos(Types.MOTION_BLOCKING_NO_LEAVES, context.origin());
    Axis axis = context.random().nextBoolean() ? Axis.X : Axis.Z;
    StarGateStructure starGateStructure = StarGateStructure.INSTANCE;
    LevelAccessor levelAccessor = (LevelAccessor) worldGenLevel;
    if (!starGateStructure.canPlace(levelAccessor, blockPos, axis)) {
      return false;
    }

    StarGateStructure.INSTANCE.place((LevelAccessor) worldGenLevel, blockPos, axis, false);
    return true;
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/dimension/StarGateFeature.kt
package com.cxxcxx.zinecraft.core.dimension

import com.mojang.serialization.Codec
import net.minecraft.core.Direction
import net.minecraft.world.level.Level
import net.minecraft.world.level.levelgen.Heightmap
import net.minecraft.world.level.levelgen.feature.Feature
import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext
import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration

/** 主世界雪原自然生成的星门入口。 */
class StarGateFeature : Feature<NoneFeatureConfiguration>(NoneFeatureConfiguration.CODEC) {
  override fun place(context: FeaturePlaceContext<NoneFeatureConfiguration>): Boolean {
    val level = context.level()
    if (level.level.dimension() != Level.OVERWORLD) return false

    val base = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, context.origin())
    val axis = if (context.random().nextBoolean()) Direction.Axis.X else Direction.Axis.Z
    if (!StarGateStructure.canPlace(level, base, axis)) return false
    StarGateStructure.place(level, base, axis)
    return true
  }
}
